crates/biome_js_analyze/src/lint/nursery/no_restricted_types.rs

```diff
@@ -0,0 +1,356 @@
+use std::fmt::Display;
+
+use ::serde::{Deserialize, Serialize};
+use biome_analyze::context::RuleContext;
+use biome_analyze::{declare_lint_rule, ActionCategory, FixKind, Rule, RuleDiagnostic, RuleSource};
+use biome_console::markup;
+use biome_deserialize_macros::Deserializable;
+use biome_js_factory::make;
+use biome_js_syntax::{
+    JsReferenceIdentifier, TextRange, TsIntersectionTypeElementList, TsObjectType, TsReferenceType,
+    TsTypeConstraintClause,
+};
+use biome_rowan::{declare_node_union, AstNode, AstNodeList, BatchMutationExt};
+use rustc_hash::FxHashMap;
+
+use crate::services::semantic::Semantic;
+use crate::JsRuleAction;
+
+#[cfg(feature = "schemars")]
+use schemars::JsonSchema;
+
+declare_lint_rule! {
+    /// Disallow primitive type aliases, misleading or user defined types.
+    ///
+    /// - Enforce consistent names for primitive types
+    ///
+    ///   Primitive types have aliases.
+    ///   For example, `Number` is an alias of `number`.
+    ///   The rule recommends the lowercase primitive type names.
+    ///
+    /// - Disallow the `Function` type
+    ///
+    ///   The `Function` type is loosely typed and is thus considered dangerous or harmful.
+    ///   `Function` is equivalent to the type `(...rest: any[]) => any` that uses the unsafe `any` type.
+    ///
+    /// - Disallow the misleading non-nullable type `{}`
+    ///
+    ///   In TypeScript, the type `{}` doesn't represent an empty object.
+    ///   It represents any value except `null` and `undefined`.
+    ///   The following TypeScript example is perfectly valid:
+    ///
+    ///   ```ts,expect_diagnostic
+    ///   const n: {} = 0
+    ///   ```
+    ///
+    ///   To represent an empty object, you should use `{ [k: string]: never }` or `Record<string, never>`.
+    ///
+    ///   To avoid any confusion, the rule forbids the use of the type `{}`, except in two situations:
+    ///
+    ///   1. In type constraints to restrict a generic type to non-nullable types:
+    ///
+    ///   ```ts
+    ///   function f<T extends {}>(x: T) {
+    ///       assert(x != null);
+    ///   }
+    ///   ```
+    ///
+    ///   2. In a type intersection to narrow a type to its non-nullable equivalent type:
+    ///
+    ///   ```ts
+    ///   type NonNullableMyType = MyType & {};
+    ///   ```
+    ///
+    ///   In this last case, you can also use the `NonNullable` utility type:
+    ///
+    ///   ```ts
+    ///   type NonNullableMyType = NonNullable<MyType>;
+    ///   ```
+    ///
```

Comment:
We should add an `Options` section (See rules with options for some examples, e.g. `useNamingCOnvention`, `useFilenamingConvention`).